### DISCUSSION TITLE
"Unexpected resource" error in Request Forwarding LSP

### DISCUSSION THREAD

**Discussion by SebasF1349** [microsoft/vscode-discussions#310] [Extension Development QnA] [answered]
I've made a VSCode LSP for a new "language" following the guides and samples. The "language" is a mix of html, js and a special formatting, so I'm using Request Forwarding to get html and js completion and hover.

You can find the repo here: https://github.com/SebasF1349/SAMMI-sef-VSCode

I have an issue with js completion as sometimes doesn't work or only works after many minutes. But it's not always the case, looks random (I would say 80% of the time doesn't work, but 20% works perfectly).

Here is the (I think) relevant code (full file: https://github.com/SebasF1349/SAMMI-sef-VSCode/blob/master/client/src/extension.ts):

```typescript
provideCompletionItem: async (document, position, context, token, next) => {
    const region = fileRegion(document, position, true);
    if (region === "SAMMI") {
        return await next(document, position, context, token);
    }

    const originalUri = document.uri.toString(true);
    virtualDocumentContents.set(originalUri, getVirtualContent(region, document.getText()));

    const vdocUriString = `embedded-content://${region}/${encodeURIComponent(originalUri)}.${region}`;
    const vdocUri = Uri.parse(vdocUriString);
    return await commands.executeCommand<CompletionList>(
        "vscode.executeCompletionItemProvider",
        vdocUri,
        position,
        context.triggerCharacter
    );
}
```

When `region` is equal to `html`, it works 100% of the time. And getVirtualContent works exactly the same for js and html (https://github.com/SebasF1349/SAMMI-sef-VSCode/blob/master/client/src/embeddedSupport.ts) so I'm very confused.

Also, hover works 100% (in html and in js), and the code is mostly the same.

When debugging, *sometimes* (but not every time completion fails) I get this error in the terminal:

`Unexpected resource embedded-content://js/file%3A///c%3A/file/path/to/testing/file.sef.js`
in the file `c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:98`

Where I'm getting that error? Any have any idea what I'm doing wrong?

**Comment by SebasF1349** [comment]
I think I found something.

If after opening the file of the new language a hover over some js function (before writing anything), js/ts language features get initialised and after that the js completion works most of the time, but still not always (something like 60% of the time works now). And sometimes it doesn't work but if I hover over a variable and get its type, then I get autocomplete on that variable.

So looks like VSCode needs to initialise JS before I can forward my file, and that looks very annoying. It gets initializated on hover but not on completion? Can I control that anyway from the extension? I'm very confused, but even then it doesn't work always.

**Comment by isidorn** [comment]
@dbaeumer might be able to help here

**Comment by dbaeumer** [comment]
@mjbvz might be the better person to answer. To my knowledge there is currently no API to find out when the JS/TS language service is ready to use.

**Comment by mjbvz** [comment]
Yes there is no api for this. WIth JS/TS, we return fast results for completions and hovers while the project is initializing. We only start returning the full results after the project has been loaded. But extensions cannot tell which result they get back from `vscode.executeCompletionItemProvider` and similar commands

**Comment by SebasF1349** [comment]
So is there a workaround? Like some way to tell VSCode to initializate JS even before the user does anything. I tried to send a dummy line to `vscode.executeHoverProvider` at start, but fail to make it work.

Or we can't just forward content to the JS LSP?

**Comment by mjbvz** [comment, marked as answer]
Oh actually this message may not be about anything LSP related. Instead you are likely hitting this code:

https://github.com/microsoft/vscode/blob/356b57c157a39ef957bca466240711e682ccca54/extensions/typescript-language-features/src/typescriptServiceClient.ts#L697

This suggests that you are making requests about a document that is not actually opened in VS Code (or one that was opened previously and has now been closed). Our JS/TS support assumes that we only get intellisense requests for opened documents. You may be able to workaround this by calling `vscode.workspace.openTextDocument` before making the request

**Comment by SebasF1349** [comment]
That looks to be *partially* correct.

Using `vscode.workspace.openTextDocument` to open the file (with extension .sef) looks to make the autocomplete always work, so that's a huge advance.

But I still receive the error `Unexpected resource embedded-content://js/file%3A///c%3A/file/path/to/testing/file.sef.js`. I don't think I mentioned earlier, but the error shows when triggering the autocomplete (with `.`) and after adding the suggestion (pressing tab in the method).

So apparently the error was not related to autocomplete not showing, but it's related to that part of the code as it doesn't show with any other action (not html autocomplete, not hover or diagnostics). I will have to keep investigating that.

I'm not entirely happy I still get an error, but at least it seems to be working.

Thanks a lot.

**Reply by burkeholland** [reply]
Marking @mjbvz's answer as correct. If you feel it's not correct _enough_, please feel free to re-open as unanswered.
